File: benchmarks/go/bench.go
// Command bench runs the Go faster-whisper port over the shared benchmark
// config and writes results in the same JSON schema as the Python baseline
// (see ../python/bench.py), so the two can be diffed with ../compare.py.
package main

import (
	"context"
	"encoding/binary"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"math"
	"os"
	"path/filepath"
	"strings"
	"time"

	whisper "github.com/timohahaa/faster-whisper-go"
)

const sampleRate = 16000

type fileSpec struct {
	Path     string `json:"path"`
	Language string `json:"language"`
}

type vadSpec struct {
	Threshold            float32 `json:"threshold"`
	MinSpeechDurationMs  int     `json:"min_speech_duration_ms"`
	MaxSpeechDurationS   float64 `json:"max_speech_duration_s"`
	MinSilenceDurationMs int     `json:"min_silence_duration_ms"`
	SpeechPadMs          int     `json:"speech_pad_ms"`
}

type config struct {
	Model       string `json:"model"`
	Device      string `json:"device"`
	DeviceIndex int    `json:"device_index"`
	ComputeType string `json:"compute_type"`
	CPUThreads  int    `json:"cpu_threads"`
	NumWorkers  int    `json:"num_workers"`

	BatchSize      int  `json:"batch_size"`
	WordTimestamps bool `json:"word_timestamps"`
	VadFilter      bool `json:"vad_filter"`
	Warmup         bool `json:"warmup"`

	BeamSize                  int       `json:"beam_size"`
	BestOf                    int       `json:"best_of"`
	Temperature               []float32 `json:"temperature"`
	CompressionRatioThreshold float32   `json:"compression_ratio_threshold"`
	LogProbThreshold          float32   `json:"log_prob_threshold"`
	NoSpeechThreshold         float32   `json:"no_speech_threshold"`
	ConditionOnPreviousText   bool      `json:"condition_on_previous_text"`

	Vad vadSpec `json:"vad"`

	Datadir string     `json:"datadir"`
	Files   []fileSpec `json:"files"`
}

type wordOut struct {
	Start       float64 `json:"start"`
	End         float64 `json:"end"`
	Word        string  `json:"word"`
	Probability float64 `json:"probability"`
}

type segmentOut struct {
	ID    int       `json:"id"`
	Start float64   `json:"start"`
	End   float64   `json:"end"`
	Text  string    `json:"text"`
	Words []wordOut `json:"words"`
}

type fileResult struct {
	File                string       `json:"file"`
	LanguageRequested   string       `json:"language_requested"`
	LanguageDetected    string       `json:"language_detected"`
	LanguageProbability float64      `json:"language_probability"`
	AudioDurationSec    float64      `json:"audio_duration_sec"`
	TranscribeTimeSec   float64      `json:"transcribe_time_sec"`
	RTF                 float64      `json:"rtf"`
	SpeedupVsRealtime   float64      `json:"speedup_vs_realtime"`
	NumSegments         int          `json:"num_segments"`
	NumWords            int          `json:"num_words"`
	Text                string       `json:"text"`
	Segments            []segmentOut `json:"segments"`
}

type payload struct {
	Implementation string       `json:"implementation"`
	Mode           string       `json:"mode"`
	Model          string       `json:"model"`
	Device         string       `json:"device"`
	DeviceIndex    int          `json:"device_index"`
	ComputeType    string       `json:"compute_type"`
	BatchSize      int          `json:"batch_size"`
	WordTimestamps bool         `json:"word_timestamps"`
	VadFilter      bool         `json:"vad_filter"`
	ModelLoadSec   float64      `json:"model_load_sec"`
	Results        []fileResult `json:"results"`
}

func main() {
	defaultConfig := "../config.json"
	if _, err := os.Stat(defaultConfig); err != nil {
		defaultConfig = "config.json"
	}
	configPath := flag.String("config", defaultConfig, "path to shared benchmark config.json")
	mode := flag.String("mode", "both", "batched | sequential | both")
	outDir := flag.String("out", "../results", "output directory for result JSON files")
	filesFilter := flag.String("files", "", "comma-separated list of file names to run (default: all)")
	flag.Parse()

	cfg, cfgDir, err := loadConfig(*configPath)
	if err != nil {
		fail("load config: %v", err)
	}

	datadir := cfg.Datadir
	if datadir == "" {
		datadir = "../.testdata"
	}
	if !filepath.IsAbs(datadir) {
		datadir = filepath.Join(cfgDir, datadir)
	}

	if err := os.MkdirAll(*outDir, 0o755); err != nil {
		fail("create out dir: %v", err)
	}

	files := cfg.Files
	if *filesFilter != "" {
		wanted := map[string]bool{}
		for _, w := range strings.Split(*filesFilter, ",") {
			wanted[strings.TrimSpace(w)] = true
		}
		var kept []fileSpec
		for _, f := range files {
			if wanted[f.Path] || wanted[filepath.Base(f.Path)] {
				kept = append(kept, f)
			}
		}
		files = kept
	}

	var modes []string
	switch *mode {
	case "both":
		modes = []string{"batched", "sequential"}
	default:
		modes = []string{*mode}
	}

	fmt.Printf("loading model=%s device=%s compute_type=%s ...\n", cfg.Model, cfg.Device, cfg.ComputeType)
	loadStart := time.Now()
	model, err := whisper.Load(cfg.Model, whisper.ModelConfig{
		Device:      cfg.Device,
		ComputeType: cfg.ComputeType,
		DeviceIndex: []int{cfg.DeviceIndex},
		CPUThreads:  cfg.CPUThreads,
		NumWorkers:  cfg.NumWorkers,
	})
	if err != nil {
		fail("load model: %v", err)
	}
	defer model.Close()
	modelLoadSec := time.Since(loadStart).Seconds()
	fmt.Printf("model loaded in %.2fs\n", modelLoadSec)

	// Read every WAV once up front so file IO is excluded from the timed region.
	decoded := map[string][]float32{}
	for _, f := range files {
		p := filepath.Join(datadir, f.Path)
		samples, err := readWAV(p)
		if err != nil {
			fmt.Printf("  WARN: cannot read %s: %v, skipping\n", p, err)
			continue
		}
		decoded[f.Path] = samples
	}

	ctx := context.Background()

	for _, m := range modes {
		if cfg.Warmup && len(decoded) > 0 {
			var warm []float32
			for _, f := range files {
				if s, ok := decoded[f.Path]; ok {
					warm = s
					break
				}
			}
			if n := sampleRate * 30; len(warm) > n {
				warm = warm[:n]
			}
			fmt.Printf("[%s] warmup ...\n", m)
			_, _ = transcribe(ctx, model, m, warm, "en", cfg)
		}

		var results []fileResult
		for _, f := range files {
			samples, ok := decoded[f.Path]
			if !ok {
				continue
			}
			audioDur := float64(len(samples)) / sampleRate
			fmt.Printf("[%s] transcribing %s (%.1fs) ...\n", m, f.Path, audioDur)

			start := time.Now()
			res, err := transcribe(ctx, model, m, samples, f.Language, cfg)
			elapsed := time.Since(start).Seconds()
			if err != nil {
				fail("transcribe %s: %v", f.Path, err)
			}

			fr := buildFileResult(f, res, audioDur, elapsed)
			results = append(results, fr)
			fmt.Printf("    -> %.2fs (%.1fx realtime, lang=%s, segments=%d)\n",
				fr.TranscribeTimeSec, fr.SpeedupVsRealtime, fr.LanguageDetected, fr.NumSegments)
		}

		out := payload{
			Implementation: "go",
			Mode:           m,
			Model:          cfg.Model,
			Device:         cfg.Device,
			DeviceIndex:    cfg.DeviceIndex,
			ComputeType:    cfg.ComputeType,
			BatchSize:      cfg.BatchSize,
			WordTimestamps: cfg.WordTimestamps,
			VadFilter:      cfg.VadFilter,
			ModelLoadSec:   round(modelLoadSec, 3),
			Results:        results,
		}
		outPath := filepath.Join(*outDir, fmt.Sprintf("go_%s.json", m))
		if err := writeJSON(outPath, out); err != nil {
			fail("write %s: %v", outPath, err)
		}
		fmt.Printf("[%s] wrote %s\n", m, outPath)
	}
}

func transcribe(ctx context.Context, model *whisper.Model, mode string, samples []float32, language string, cfg config) (*whisper.Result, error) {
	tc := whisper.TranscribeConfig{
		Language:                  language,
		BeamSize:                  cfg.BeamSize,
		BestOf:                    cfg.BestOf,
		Temperature:               cfg.Temperature,
		CompressionRatioThreshold: f32ptr(cfg.CompressionRatioThreshold),
		LogProbThreshold:          f32ptr(cfg.LogProbThreshold),
		NoSpeechThreshold:         f32ptr(cfg.NoSpeechThreshold),
		ConditionOnPreviousText:   boolptr(cfg.ConditionOnPreviousText),
		WordTimestamps:            cfg.WordTimestamps,
		VadFilter:                 cfg.VadFilter,
		VadConfig: &whisper.VadConfig{
			Threshold:            cfg.Vad.Threshold,
			MinSpeechDurationMs:  cfg.Vad.MinSpeechDurationMs,
			MaxSpeechDurationS:   cfg.Vad.MaxSpeechDurationS,
			MinSilenceDurationMs: cfg.Vad.MinSilenceDurationMs,
			SpeechPadMs:          cfg.Vad.SpeechPadMs,
		},
		BatchSize: cfg.BatchSize,
	}
	if mode == "batched" {
		return model.TranscribeBatched(ctx, samples, tc)
	}
	return model.Transcribe(ctx, samples, tc)
}

func buildFileResult(f fileSpec, res *whisper.Result, audioDur, elapsed float64) fileResult {
	segs := make([]segmentOut, 0, len(res.Segments))
	numWords := 0
	var textParts []string
	for _, s := range res.Segments {
		var words []wordOut
		for _, w := range s.Words {
			words = append(words, wordOut{
				Start:       round(w.Start.Seconds(), 3),
				End:         round(w.End.Seconds(), 3),
				Word:        w.Word,
				Probability: round(float64(w.Probability), 4),
			})
		}
		numWords += len(words)
		segs = append(segs, segmentOut{
			ID:    s.ID,
			Start: round(s.Start.Seconds(), 3),
			End:   round(s.End.Seconds(), 3),
			Text:  s.Text,
			Words: words,
		})
		textParts = append(textParts, strings.TrimSpace(s.Text))
	}

	rtf := 0.0
	speedup := 0.0
	if audioDur > 0 {
		rtf = elapsed / audioDur
	}
	if elapsed > 0 {
		speedup = audioDur / elapsed
	}

	return fileResult{
		File:                f.Path,
		LanguageRequested:   f.Language,
		LanguageDetected:    res.Info.Language,
		LanguageProbability: round(float64(res.Info.LanguageProbability), 4),
		AudioDurationSec:    round(audioDur, 3),
		TranscribeTimeSec:   round(elapsed, 3),
		RTF:                 round(rtf, 4),
		SpeedupVsRealtime:   round(speedup, 2),
		NumSegments:         len(segs),
		NumWords:            numWords,
		Text:                strings.TrimSpace(strings.Join(textParts, " ")),
		Segments:            segs,
	}
}

func loadConfig(path string) (config, string, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return config{}, "", err
	}
	var cfg config
	if err := json.Unmarshal(data, &cfg); err != nil {
		return config{}, "", err
	}
	abs, err := filepath.Abs(path)
	if err != nil {
		return config{}, "", err
	}
	return cfg, filepath.Dir(abs), nil
}

func writeJSON(path string, v any) error {
	data, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o644)
}

// round rounds half away from zero, matching the Python baseline's rnd().
func round(v float64, places int) float64 {
	p := math.Pow(10, float64(places))
	return math.Round(v*p) / p
}

func f32ptr(v float32) *float32 { return &v }
func boolptr(v bool) *bool      { return &v }

func fail(format string, args ...any) {
	fmt.Fprintf(os.Stderr, "error: "+format+"\n", args...)
	os.Exit(1)
}

// readWAV reads a 16-bit PCM mono 16kHz WAV file and returns float32 samples in
// [-1, 1). Input must already be in this format.
func readWAV(path string) ([]float32, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var riffHeader struct {
		ID   [4]byte
		Size uint32
		Type [4]byte
	}
	if err := binary.Read(f, binary.LittleEndian, &riffHeader); err != nil {
		return nil, fmt.Errorf("read RIFF header: %w", err)
	}
	if string(riffHeader.ID[:]) != "RIFF" || string(riffHeader.Type[:]) != "WAVE" {
		return nil, fmt.Errorf("not a RIFF/WAVE file")
	}

	var (
		fmtFound  bool
		audioFmt  uint16
		nChannels uint16
		sampleR   uint32
		bps       uint16
	)

	for {
		var chunkID [4]byte
		var chunkSize uint32
		if err := binary.Read(f, binary.LittleEndian, &chunkID); err != nil {
			if err == io.EOF {
				break
			}
			return nil, fmt.Errorf("read chunk ID: %w", err)
		}
		if err := binary.Read(f, binary.LittleEndian, &chunkSize); err != nil {
			return nil, fmt.Errorf("read chunk size: %w", err)
		}

		switch string(chunkID[:]) {
		case "fmt ":
			var fmtChunk struct {
				AudioFormat   uint16
				NumChannels   uint16
				SampleRate    uint32
				ByteRate      uint32
				BlockAlign    uint16
				BitsPerSample uint16
			}
			if err := binary.Read(f, binary.LittleEndian, &fmtChunk); err != nil {
				return nil, fmt.Errorf("read fmt chunk: %w", err)
			}
			audioFmt = fmtChunk.AudioFormat
			nChannels = fmtChunk.NumChannels
			sampleR = fmtChunk.SampleRate
			bps = fmtChunk.BitsPerSample
			fmtFound = true
			if extra := int64(chunkSize) - 16; extra > 0 {
				if _, err := f.Seek(extra, io.SeekCurrent); err != nil {
					return nil, fmt.Errorf("skip fmt extra bytes: %w", err)
				}
			}

		case "data":
			if !fmtFound {
				return nil, fmt.Errorf("data chunk before fmt chunk")
			}
			if audioFmt != 1 {
				return nil, fmt.Errorf("unsupported audio format %d (want PCM=1)", audioFmt)
			}
			if nChannels != 1 {
				return nil, fmt.Errorf("unsupported channels %d (want mono=1)", nChannels)
			}
			if sampleR != sampleRate {
				return nil, fmt.Errorf("unsupported sample rate %d (want %d)", sampleR, sampleRate)
			}
			if bps != 16 {
				return nil, fmt.Errorf("unsupported bits per sample %d (want 16)", bps)
			}
			nSamples := int(chunkSize) / 2
			raw := make([]int16, nSamples)
			if err := binary.Read(f, binary.LittleEndian, raw); err != nil {
				return nil, fmt.Errorf("read PCM data: %w", err)
			}
			samples := make([]float32, nSamples)
			for i, s := range raw {
				samples[i] = float32(s) / 32768.0
			}
			return samples, nil

		default:
			if _, err := f.Seek(int64(chunkSize), io.SeekCurrent); err != nil {
				return nil, fmt.Errorf("skip chunk %q: %w", chunkID, err)
			}
		}
	}

	return nil, fmt.Errorf("no data chunk found")
}

File: benchmarks/python/bench.py
#!/usr/bin/env python3
"""Benchmark the reference Python faster-whisper implementation.

Two pipelines: "batched" (BatchedInferencePipeline) and "sequential"
(WhisperModel.transcribe). Every decode and VAD parameter is read from the
shared config.json and passed explicitly, so this baseline and the Go port (see
../go/bench.go) run with identical settings. Results are written to JSON for
comparison with compare.py.
"""

import argparse
import json
import math
import time
import wave
from pathlib import Path

import numpy as np
from faster_whisper import WhisperModel, BatchedInferencePipeline
from faster_whisper.vad import VadOptions

SAMPLE_RATE = 16000


def rnd(v: float, places: int) -> float:
    """Round half away from zero, identical to the Go port's round()."""
    p = 10**places
    return math.copysign(math.floor(abs(v) * p + 0.5), v) / p


def load_config(path: Path) -> dict:
    with open(path, encoding="utf-8") as f:
        return json.load(f)


def resolve_datadir(cfg: dict, config_path: Path) -> Path:
    datadir = Path(cfg.get("datadir", "../.testdata"))
    if not datadir.is_absolute():
        datadir = (config_path.parent / datadir).resolve()
    return datadir


def read_wav(path: Path) -> np.ndarray:
    """Read a 16 kHz mono 16-bit PCM WAV into float32 samples in [-1, 1)."""
    with wave.open(str(path), "rb") as w:
        if w.getframerate() != SAMPLE_RATE:
            raise ValueError(f"{path}: sample rate {w.getframerate()} != {SAMPLE_RATE}")
        if w.getnchannels() != 1:
            raise ValueError(f"{path}: channels {w.getnchannels()} != 1 (mono)")
        if w.getsampwidth() != 2:
            raise ValueError(f"{path}: sample width {w.getsampwidth()*8} bits != 16")
        raw = w.readframes(w.getnframes())
    return np.frombuffer(raw, dtype="<i2").astype(np.float32) / 32768.0


def vad_parameters(cfg: dict) -> VadOptions:
    vad = cfg.get("vad", {})
    max_speech = float(vad.get("max_speech_duration_s", 0))
    return VadOptions(
        threshold=float(vad.get("threshold", 0.5)),
        min_speech_duration_ms=int(vad.get("min_speech_duration_ms", 0)),
        max_speech_duration_s=max_speech if max_speech > 0 else float("inf"),
        min_silence_duration_ms=int(vad.get("min_silence_duration_ms", 2000)),
        speech_pad_ms=int(vad.get("speech_pad_ms", 400)),
    )


def segments_to_dicts(segments) -> list[dict]:
    # Consuming the generator is what triggers the actual transcription work.
    out = []
    for seg in segments:
        words = None
        if seg.words:
            words = [
                {
                    "start": rnd(w.start, 3),
                    "end": rnd(w.end, 3),
                    "word": w.word,
                    "probability": rnd(w.probability, 4),
                }
                for w in seg.words
            ]
        out.append(
            {
                "id": seg.id,
                "start": rnd(seg.start, 3),
                "end": rnd(seg.end, 3),
                "text": seg.text,
                "words": words,
            }
        )
    return out


def run_one(model, batched, mode: str, audio, language: str, cfg: dict) -> dict:
    common = dict(
        language=language or None,
        task="transcribe",
        beam_size=int(cfg.get("beam_size", 5)),
        best_of=int(cfg.get("best_of", 5)),
        temperature=cfg.get("temperature", [0.0, 0.2, 0.4, 0.6, 0.8, 1.0]),
        compression_ratio_threshold=cfg.get("compression_ratio_threshold", 2.4),
        log_prob_threshold=cfg.get("log_prob_threshold", -1.0),
        no_speech_threshold=cfg.get("no_speech_threshold", 0.6),
        condition_on_previous_text=bool(cfg.get("condition_on_previous_text", True)),
        word_timestamps=bool(cfg.get("word_timestamps", True)),
        vad_filter=bool(cfg.get("vad_filter", True)),
        vad_parameters=vad_parameters(cfg),
    )

    audio_duration = len(audio) / SAMPLE_RATE

    start = time.perf_counter()
    if mode == "batched":
        segments, info = batched.transcribe(
            audio, batch_size=int(cfg.get("batch_size", 8)), **common
        )
    else:
        segments, info = model.transcribe(audio, **common)
    seg_dicts = segments_to_dicts(segments)
    elapsed = time.perf_counter() - start

    text = " ".join(s["text"].strip() for s in seg_dicts).strip()
    num_words = sum(len(s["words"]) for s in seg_dicts if s["words"])

    return {
        "language_detected": info.language,
        "language_probability": rnd(info.language_probability, 4),
        "audio_duration_sec": rnd(audio_duration, 3),
        "transcribe_time_sec": rnd(elapsed, 3),
        "rtf": rnd(elapsed / audio_duration, 4) if audio_duration else 0.0,
        "speedup_vs_realtime": rnd(audio_duration / elapsed, 2) if elapsed else 0.0,
        "num_segments": len(seg_dicts),
        "num_words": num_words,
        "text": text,
        "segments": seg_dicts,
    }


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--config", default=str(Path(__file__).parent.parent / "config.json"))
    ap.add_argument("--mode", choices=["batched", "sequential", "both"], default="both")
    ap.add_argument("--out", default=str(Path(__file__).parent.parent / "results"))
    ap.add_argument("--files", nargs="*", help="only run these file names (basename match)")
    args = ap.parse_args()

    config_path = Path(args.config).resolve()
    cfg = load_config(config_path)
    datadir = resolve_datadir(cfg, config_path)
    out_dir = Path(args.out).resolve()
    out_dir.mkdir(parents=True, exist_ok=True)

    files = cfg["files"]
    if args.files:
        wanted = set(args.files)
        files = [f for f in files if Path(f["path"]).name in wanted or f["path"] in wanted]

    modes = ["batched", "sequential"] if args.mode == "both" else [args.mode]

    print(
        f"loading model={cfg['model']} device={cfg['device']} "
        f"compute_type={cfg['compute_type']} ..."
    )
    load_start = time.perf_counter()
    model = WhisperModel(
        cfg["model"],
        device=cfg.get("device", "cpu"),
        device_index=cfg.get("device_index", 0),
        compute_type=cfg.get("compute_type", "int8"),
        cpu_threads=int(cfg.get("cpu_threads", 0)),
        num_workers=int(cfg.get("num_workers", 1)),
    )
    batched = BatchedInferencePipeline(model=model)
    model_load_sec = time.perf_counter() - load_start
    print(f"model loaded in {model_load_sec:.2f}s")

    # Read every WAV once up front so file IO is excluded from the timed region.
    decoded: dict[str, np.ndarray] = {}
    for f in files:
        p = datadir / f["path"]
        if not p.exists():
            print(f"  WARN: missing {p}, skipping")
            continue
        decoded[f["path"]] = read_wav(p)

    for mode in modes:
        if cfg.get("warmup") and decoded:
            first = next(iter(decoded.values()))
            warm = first[: SAMPLE_RATE * 30]
            print(f"[{mode}] warmup ...")
            run_one(model, batched, mode, warm, "en", cfg)

        results = []
        for f in files:
            if f["path"] not in decoded:
                continue
            audio = decoded[f["path"]]
            print(f"[{mode}] transcribing {f['path']} ({len(audio)/SAMPLE_RATE:.1f}s) ...")
            r = run_one(model, batched, mode, audio, f.get("language", ""), cfg)
            r["file"] = f["path"]
            r["language_requested"] = f.get("language", "")
            results.append(r)
            print(
                f"    -> {r['transcribe_time_sec']:.2f}s "
                f"({r['speedup_vs_realtime']:.1f}x realtime, "
                f"lang={r['language_detected']}, segments={r['num_segments']})"
            )

        payload = {
            "implementation": "python",
            "mode": mode,
            "model": cfg["model"],
            "device": cfg.get("device"),
            "device_index": cfg.get("device_index", 0),
            "compute_type": cfg.get("compute_type"),
            "batch_size": cfg.get("batch_size", 8),
            "word_timestamps": cfg.get("word_timestamps", True),
            "vad_filter": cfg.get("vad_filter", True),
            "model_load_sec": rnd(model_load_sec, 3),
            "results": results,
        }
        out_path = out_dir / f"python_{mode}.json"
        with open(out_path, "w", encoding="utf-8") as fh:
            json.dump(payload, fh, ensure_ascii=False, indent=2)
        print(f"[{mode}] wrote {out_path}")


if __name__ == "__main__":
    main()
